Make annotations great again!

denaitre opened the pull request:
1. Add documentation.
2. An event handles now a 'title', a 'text' and 'tags':
    * title defines the event type (playbook, play, task);
    * text defines the event name;
    * tags contain further information about the event (user, node, kolla-related, task execution time).
3. Because recording every task saved too many useless information, tasks are
no longer recorded when they begin. In order to record only the tasks that
imply a modification (marked as 'changed' in Ansible), this commit checks the
status of tasks before adding them to InfluxDB.
4. InfluxDB connexion is tested, if not working, the plugin is disabled
and a warning message appears.
5. The configuration is provided to `kolla-ansible` to pass `influx_vip`
which is required to record events during the `os_install` phase, as well as
different information that could be used as tags.

Related to #11

Result:
![annotations8_crop](https://cloud.githubusercontent.com/assets/1255977/24043621/c12c008e-0b17-11e7-99e6-84d25b10db42.png)

Potential future step:
* Get annotations inside rally benchs.
* Other tags?

msimonin commented on enos/ansible/plugins/callback/influxdb_events.py:
```diff
@@ -30,107 +26,129 @@ class CallbackModule(CallbackBase):
+        # Check if `enable_monitoring` is set
+        if not self.extra_vars.get('enable_monitoring', True):
+            # Disable the plugin if InfluxDB is unreachable
+            self.disabled = True
```
Is `self.disabled` a pre-defined attribute of ansible Callback plugins ? Otherwise It seems unused in this file.

denaitre replied:
It is.

msimonin commented on enos/ansible/plugins/callback/influxdb_events.py:
```diff
@@ -30,107 +26,129 @@ class CallbackModule(CallbackBase):
+
         try:
-            self.influxdb = InfluxDBClient(_host, _port, _user, _pass, _dbname)
+            self.influxdb.write_points(self.events, time_precision='u')
```
Several questions here, in case influxdb isn't reachable, 

* What will be the effect of setting `disabled=True` ? 
* I'm afraid it will block Ansible until a timeout occurs. If it's the case we should maybe make a socket test prior to sending the actual datas (Maybe the client has a built-in test ?).

denaitre replied:
We check `enable_monitoring` in the function `v2_playbook_on_start` (l.74) which should prevent timeouts due to a missing InfluxDB. Another (flawed) solution would be to set the time before timeout [1].

[1] https://github.com/influxdata/influxdb-python/blob/master/influxdb/client.py#L53

msimonin replied:
ok, I see.

msimonin commented on enos/enos.py:
```diff
@@ -219,9 +219,13 @@ def install_os(env=None, **kwargs):
                       "--passwords", "%s/passwords.yml" % env['resultdir'],
-                      "--configdir", "%s" % env['resultdir']])
+                      "--configdir", "%s" % env['resultdir'],
+                      "--extra", "%s" % str(env['config']).replace(" ","")])
```
Got it we make a system call so we need to escape spaces in the command line.

msimonin replied:
we could add the specific vars in `generated_kolla_vars` and remove the need for the `--extra` option

denaitre commented:
@rcherrueau @msimonin Double done! Thanks for the reviews.

Here's a screen with event tag and openstack release.
![annotations9_crop](https://cloud.githubusercontent.com/assets/1255977/24146990/84e2fde8-0e38-11e7-9945-08909b9d8c2d.png)

msimonin commented on enos/enos.py:
```diff
@@ -187,7 +187,12 @@ def install_os(env=None, **kwargs):
+        # Specific variables for the Ansible plugin
+        'influx_vip':                 env['config']['influx_vip'],
+        'kolla_ref':                  env['config']['kolla_ref'],
+        'enable_monitoring':
```
IMO we should add `enable_monitoring` only if the key exist. We shouldn't choose a default value here since it will be set by ansible automatically when reading the `groups_vars/all.yml`

msimonin replied:
This is not looking great but something like : 
```
if 'enable_monitoring' in env['config']:
  generated_kolla_vars.update({'enable_monitoring', env['config']['enable_monitoring']})

rcherrueau replied:
👍

denaitre commented:
Hi there!

@msimonin I have modified the code to check first if `enable_monitoring` is provided in `reservation.yml`, otherwise it is read from the Ansible variables (where `enable_monitoring` is provided from `all.yml`) [1].

However, during the `install_os` phase, the appropriate Ansible variables are not loaded before Ansible facts are gathered (which happens in `site.yml`)... Thus I consider the value `enable_monitoring = True` by default. If it is not the case, it will be corrected in the next Ansible `Play` call.

Does it make sense?

By the way, I have added a new tag: the region name:

![annotations10_crop](https://cloud.githubusercontent.com/assets/1255977/24323415/3b961a68-1175-11e7-9fc4-eb42e77731a8.png)

[1] https://github.com/BeyondTheClouds/enos/pull/93/files#diff-8c23be17cb3fb1bc6cfb9c72c566d9daR98

msimonin commented on enos/ansible/plugins/callback/influxdb_events.py:
```diff
@@ -3,134 +3,189 @@
+        # NOTE(dpertin) before gathering facts `enable_monitoring` is not
+        # provided as a playbook variable, thus we consider it is enabled by
+        # default.
+        enable_monitoring = self.extra_vars.get('enable_monitoring') \
```
So if I understand correctly, if the variable isn't ready (before the setup) we let the following tasks (after the setup) decide its value. 

* If the variable is explicitly set in the `reservation.yml` file this will be used.
* If not `hostvars` will hold the right value of `enable_monitoring` after the setup phase. 

I'm starting to think we could remove `enable_monitoring` from the extra_vars. Because 
* If `enabling_monitoring` is actually `True` the `hostvars` will hold the right value after the setup and we can choose `True` as the default.
* If `enabling_monitoring` is actually `False` we can still put `True` as the default for all tasks before the setup. The tasks after the setup will disabled the plugin when looking in the `hostvars`.

So the test would become : `enable_monitoring = host_vars.get('enable_monitoring', True)`.

### Attempt to check the behaviour : 

```
---
# in enos/ansible/playbook.yml
- name: test
  hosts: enos--1
  tasks:
    - name: setup
      setup:
    - name: debug
      debug:
        var: hostvars['enos--1'].enable_monitoring
```

Forcing false : 
```
╰─$ ansible-playbook -i current/multinode -e "enable_monitoring=false" enos/ansible/playbook.yml

PLAY [test] ********************************************************************

TASK [setup] *******************************************************************
ok: [enos--1]

TASK [setup] *******************************************************************
ok: [enos--1]

TASK [debug] *******************************************************************
ok: [enos--1] => {
    "hostvars['enos--1'].enable_monitoring": "false"
}

PLAY RECAP *********************************************************************
enos--1                    : ok=3    changed=0    unreachable=0    failed=0
```

Letting the default of `group_vars/all.yml`
```
╰─$ ansible-playbook -i current/multinode enos/ansible/playbook.yml

PLAY [test] ********************************************************************

TASK [setup] *******************************************************************
ok: [enos--1]

TASK [setup] *******************************************************************
ok: [enos--1]

TASK [debug] *******************************************************************
ok: [enos--1] => {
    "hostvars['enos--1'].enable_monitoring": true
}

PLAY RECAP *********************************************************************
enos--1                    : ok=3    changed=0    unreachable=0    failed=0
```

denaitre replied:
@msimonin I have removed `extra_vars` since information can be found in `host_vars`.

However, `enable_monitoring` is not available in `host_vars` with `kolla_ansible` when it is not set in `reservation.yml` since it is an Enos variable.

Thus, this new proposition checks if `enable_monitoring` is set in `host_vars`:
* if `enable_monitoring` is not set in `reservation.yml`, it will be considered `True`,
* if `enable_monitoring` is set in `reservation.yml`, it will be available anytime in `host_vars`.

msimonin commented on enos/ansible/plugins/callback/influxdb_events.py:
```diff
@@ -3,134 +3,189 @@
+        # Check if OpenStack release is set
+        if not self.openstack_release:
+            openstack_release = host_vars.get('openstack_release')
+            if openstack_release != 'auto':
```
Is `None` a problem here ?

msimonin commented on enos/ansible/plugins/callback/influxdb_events.py:
```diff
@@ -3,134 +3,189 @@
+                self.openstack_release = openstack_release
+        # Check if OpenStack region name is set
+        if not self.openstack_region_name:
+            self.openstack_region_name = host_vars.get('openstack_region_name')
```
Is `None` a problem here ?

msimonin commented on enos/ansible/plugins/callback/influxdb_events.py:
```diff
@@ -3,134 +3,189 @@
+        host = play._variable_manager._hostvars.keys()[0]
+        host_vars = play._variable_manager._hostvars[str(host)]
+
+        # If not set, fetch the value of `enable_monitoring` from either the
```
I guess it's the case for every variable ?

msimonin commented on enos/enos.py:
```diff
@@ -184,8 +184,15 @@ def install_os(env=None, **kwargs):
+        'kolla_ref':                  env['config']['kolla_ref']
     }
+    # Add `enable_configuring` if specified in the config file
+    if 'enable_monitoring' in env['config']:
```
See my previous comment , this should be removed if the behavior with `hostvars` is confirmed.

msimonin commented on docs/analysis/index.rst:
```diff
@@ -13,9 +13,27 @@ Setting ``enable_monitoring: true`` in the configuration file will deploy a moni
+Enos embeds an Ansible plugin to add annotations in Grafana.
+These annotations are marked points which provide rich information about events
+when hovered over. To enable this option, the plugin must be loaded and the
+monitoring stack must be deployed (i.e. ``enable_monitoring: true`` should be
```
I'm not sure if `enable_monitoring: true` should be in the configuration file (see comment below). But I may be wrong about how callbacks handle their variables.
